Mathlib/Data/Fin/Basic.lean

```diff
@@ -1,118 +1,720 @@
+
+end
+
+end Order
+
+section Add
+
+/-!
+### addition, numerals, and coercion from nat
+-/
+
+/-- Given a positive `n`, `fin.of_nat' i` is `i % n` as an element of `fin n`. -/
+def ofNat'' [NeZero n] (i : ℕ) : Fin n :=
+  ⟨i % n, mod_lt _ <| NeZero.pos n⟩
+#align fin.of_nat' Fin.ofNat''ₓ
+-- porting note: `Fin.ofNat'` conflicts with something in core (there the hypothesis is `n > 0`),
+-- so for now we make this double-prime `''`. This is also the reason for the dubious translation.
+
+theorem one_val' {n : ℕ} : (1 : Fin (n + 1)).val = 1 % (n + 1) := by
+  cases n
+  · simp only
+  · simp only [OfNat.ofNat, Fin.ofNat]
+    rw [Nat.mod_eq_of_lt, Fin.ofNat'_one, one_val]
+    exact Nat.succ_lt_succ (Nat.zero_lt_succ _)
+  -- used to be `rfl`
+#align fin.one_val Fin.one_val'ₓ
+-- porting note: this will conflict with the ad hoc port of `Fin.one_val`, for which the types
+-- don't match
+
+theorem coe_one' {n : ℕ} : ((1 : Fin (n + 1)) : ℕ) = 1 % (n + 1) :=
+  rfl
+#align fin.coe_one' Fin.coe_one'
+
+-- @[simp] -- Porting note: simp can prove this
+theorem val_one {n : ℕ} : (1 : Fin (n + 2)).val = 1 :=
+  rfl
+#align fin.val_one Fin.val_one
+
+-- @[simp] -- Porting note: simp can prove this
+theorem val_one' {n : ℕ} : ((1 : Fin (n + 2)) : ℕ) = 1 :=
+  rfl
+#align fin.coe_one Fin.val_one'
+-- porting note: syntactically the same as `Fin.val_one` now.
+
+@[simp]
+theorem mk_one : (⟨1, Nat.succ_lt_succ (Nat.succ_pos n)⟩ : Fin (n + 2)) = (1 : Fin _) :=
+  rfl
+#align fin.mk_one Fin.mk_one
 
-/- The basic structures on `Fin` are predicated on `Fin n` being nonempty.
+instance nontrivial {n : ℕ} : Nontrivial (Fin (n + 2)) where
+  exists_pair_ne := ⟨0, 1, (ne_iff_vne 0 1).mpr (by simp only [val_one', val_zero])⟩
+
+theorem nontrivial_iff_two_le : Nontrivial (Fin n) ↔ 2 ≤ n := by
+  rcases n with (_ | _ | n) <;>
+  simp [←Nat.one_eq_succ_zero, Fin.nontrivial, not_nontrivial, Nat.succ_le_iff]
+-- porting note: here and in the next lemma, had to use `←Nat.one_eq_succ_zero`.
+#align fin.nontrivial_iff_two_le Fin.nontrivial_iff_two_le
+
+theorem subsingleton_iff_le_one : Subsingleton (Fin n) ↔ n ≤ 1 := by
+  rcases n with (_ | _ | n) <;>
+  simp [IsEmpty.instSubsingleton, Unique.instSubsingleton, ←Nat.one_eq_succ_zero, not_subsingleton]
+#align fin.subsingleton_iff_le_one Fin.subsingleton_iff_le_one
+
+section Monoid
+
+--Porting note: removing `simp`, `simp` can prove it with AddCommMonoid instance
+protected theorem add_zero (k : Fin (n + 1)) : k + 0 = k := by
+  simp [eq_iff_veq, add_def, mod_eq_of_lt (is_lt k)]
+#align fin.add_zero Fin.add_zero
+
+--Porting note: removing `simp`, `simp` can prove it with AddCommMonoid instance
+protected theorem zero_add (k : Fin (n + 1)) : (0 : Fin (n + 1)) + k = k := by
+  simp [eq_iff_veq, add_def, mod_eq_of_lt (is_lt k)]
+#align fin.zero_add Fin.zero_add
+
+/- Porting note: (Algebraic structure on `Fin n`)
+The basic structures on `Fin` are predicated on `Fin n` being nonempty.
```

Comment:
Why are we refactoring fin in the port to use `[Nonempty (Fin n)]` on the algebraic structures? I don't think we should do a refactor just because the add_hoc port did.

Reply:
(or alternatively; if we need this refactor, we should do it in lean3 first)

Reply:
@digama0 what do you think? Should `UInt.lean` be refactored back to the way it is in Lean 3? Or should we go the `Nonempty` path?

Reply:
[ne_zero n] is surely better than [nonempty (fin n)]?

Reply:
Arguably not doing refactors at all in the port is the better way to go, unless they're unavoidable.

Reply:
@eric-wieser See https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Data.2EFin.2EBasic.20.28mathlib4.231084.29/near/319932669 . `Fin n` (where `n` is not syntactically of the form `n+1`) not being a ring in lean 3 is one of the main reasons that `zmod` even exists, it is a major pain point. In lean 4 we can't avoid this anymore because `UInt8` are new core types that rely essentially on having a ring instance on `Fin size` where `size` is some definition for a positive number like 256.

Reply:
In Lean 3 it only has to be definitionally of the form `n + 1`, doesn't it? So `fin 256` already has that ring structure. Is Lean4 different somehow here?

At any rate, it feels like if this is a big deal for the port, then we should refactor the instance in Lean3; otherwise, we'll have no idea what downstream code breaks as a result, until we're mid-port and it's too late to try a different approach.

Reply:
> In Lean 3 it only has to be definitionally of the form `n + 1`, doesn't it? So `fin 256` already has that ring structure. Is Lean4 different somehow here?

No, in both lean 3 and lean 4 instance search only unfolds reducibles, this is how you can have a different order instance on `A` and `OrderDual A`. Besides that it isn't possible to solve a unification problem like `size = ?n + 1` without unfolding and reducing `size`, which would be a huge issue if `size` is, say, 2^64....

> At any rate, it feels like if this is a big deal for the port, then we should refactor the instance in Lean3; otherwise, we'll have no idea what downstream code breaks as a result, until we're mid-port and it's too late to try a different approach.

Okay. I think the fallout from the refactor will not be too much since the new one is strictly more general (since `n+1` is nonzero, however we decide to spell that) and all the action is happening inside instance search, which is already not expected to be exactly the same between lean 3 and lean 4. But I'm not against backporting this, and indeed we might even be able to unify `zmod` and `fin` if we do (although I don't really want to propose that at this point).

Reply:
`zmod 0 = int` but `fin 0` is empty so I don't see how it's possible to unify them ...